pkg/addon/render.go

```diff
@@ -0,0 +1,278 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	common2 "github.com/oam-dev/kubevela/apis/core.oam.dev/common"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/apis/types"
+	cuemodel "github.com/oam-dev/kubevela/pkg/cue/model"
+	"github.com/oam-dev/kubevela/pkg/cue/model/value"
+	"github.com/oam-dev/kubevela/pkg/multicluster"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	addonutil "github.com/oam-dev/kubevela/pkg/utils/addon"
+)
+
+type addonCueTemplateRender struct {
+	addon     *InstallPackage
+	inputArgs map[string]interface{}
+}
+
+// This func can be used for addon render, supporting render app template and component.
+// Please notice the result will be stored in object parameter, so object must be a pointer type
+func (a addonCueTemplateRender) toObject(cueTemplate string, object interface{}) error {
+	bt, err := json.Marshal(a.inputArgs)
+	if err != nil {
+		return err
+	}
+	var contextFile = strings.Builder{}
+	var paramFile = cuemodel.ParameterFieldName + ": {}"
+	if string(bt) != "null" {
+		paramFile = fmt.Sprintf("%s: %s", cuemodel.ParameterFieldName, string(bt))
+	}
+	// addon metadata context
+	metadataJSON, err := json.Marshal(a.addon.Meta)
+	if err != nil {
+		return err
+	}
+	contextFile.WriteString(fmt.Sprintf("context: metadata: %s\n", string(metadataJSON)))
+	// parameter definition
+	contextFile.WriteString(paramFile + "\n")
+	// user custom parameter
+	contextFile.WriteString(a.addon.Parameters + "\n")
+
+	v, err := value.NewValue(contextFile.String(), nil, "")
+	if err != nil {
+		return err
+	}
+	out, err := v.LookupByScript(cueTemplate)
+	if err != nil {
+		return err
+	}
+	outputContent, err := out.LookupValue("output")
+	if err != nil {
+		return err
+	}
+	b, err := cueyaml.Encode(outputContent.CueValue())
+	if err != nil {
+		return err
+	}
+	return yaml.Unmarshal(b, object)
+}
+
+// generateAppFramework generate application from yaml defined by template.yaml or cue file from template.cue
+func generateAppFramework(addon *InstallPackage, parameters map[string]interface{}) (*v1beta1.Application, error) {
+	var app *v1beta1.Application
+	var err error
+	if len(addon.AppCueTemplate.Data) != 0 {
+		app, err = renderAppAccordingToCueTemplate(addon, parameters)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		app = addon.AppTemplate
+		if app == nil {
+			app = &v1beta1.Application{
+				TypeMeta: metav1.TypeMeta{APIVersion: "core.oam.dev/v1beta1", Kind: "Application"},
+				Spec: v1beta1.ApplicationSpec{
+					Components: []common2.ApplicationComponent{},
+				},
+			}
+		}
+		if app.Spec.Components == nil {
+			app.Spec.Components = []common2.ApplicationComponent{}
+		}
+	}
+
+	app.Name = addonutil.Addon2AppName(addon.Name)
+	// force override the namespace defined vela with DefaultVelaNS,this value can be modified by Env
+	app.SetNamespace(types.DefaultKubeVelaNS)
```

Comment:
I was wondering how can this value be modified with env?

Reply:
do we support `vela addon enable --namespace x` now ? It's a TODO?

Reply:
This is by design. All addon associated application in same name is more convenient to manage. But user still can change this by set `DEFAULT_VELA_NS` to change this.